Detect different kinds of storage devices attached to `sd-export-usb`

creviera opened the pull request:
## Description

Resolves #18 (see https://github.com/freedomofpress/securedrop-export/issues/18#issuecomment-549599905).

/cc @emkll

eloquence commented:
Moved back to "In development" for now. The plan of record discussed during standup today was to keep this branch alive for testing purposes, but to attempt a more complete resolution that will work for both print and export.

eloquence commented:
Even so, I can confirm that I was able to perform a successful export with this workaround in place! :tada: I then unlocked the USB in `Files` in `sd-export-usb`, shut down the VM, and tried again -- and it stopped working, again at the first step, even though the bus numbering hadn't changed. I think that's https://github.com/freedomofpress/securedrop-export/issues/12 but I did not get any passphrase error messages in the UI. Baby steps :)

creviera commented:
> I then unlocked the USB in Files in sd-export-usb, shut down the VM, and tried again -- and it stopped working, again at the first step, even though the bus numbering hadn't changed. I think that's #12 but I did not get any passphrase error messages in the UI. Baby steps :)

When you unlocked the USB in Files in sd-export-usb, did you tell Qubes to remember your passphrase forever? The workaround for #12 is to unplug and plug your transfer device back in. Not the best, but it has worked consistently for me.

As far as this workaround goes, I don't see why we should wait on merging a fix for a known issue, unless I'm underestimating the effort it takes to build and deploy.

emkll commented on securedrop_export/export.py:
```diff
@@ -134,27 +134,12 @@ def extract_tarball(self):
+        # If the USB is not attached via qvm-usb attach, lsblk will have a non-zero return code
         try:
-          p = subprocess.check_output(["lsusb", "-s", "{}:".format(self.pci_bus_id)])
+            subprocess.check_output(["lsblk", "-p", "-o", "KNAME", DEVICE], stderr=subprocess.PIPE)
```
thanks @creviera this does indeed work reliably for disks!

The reason we used lsusb here is because we wanted to `check_usb_connected` function to be able to detect both printers and block devices.

The command here will return  32: `lsblk: /dev/sda1/ is not a block device`. I see 2 options:

a) search for a more reliable solution that works for printers and drives
b) we can separate the preflight checks for USB connected:
    i. For disks, using the method described here
    ii. For printers, using either lsusb or another method

what do you think?

creviera replied:
Oops, somehow I missed this comment! I seeeeeee. My vote goes to whatever stabilizes export the quickest, so I think that's b, and then we can research a.

creviera commented:
> a) search for a more reliable solution that works for printers and drives

sounds like this is the path the group is going down and @rmol has already worked on some useful code for getting `busnum` and `devnum` etc. which we can use to create a more generic solution for checking whether or not a usb disk drive OR a usb-connected printer is attached. Here is the snippet of code @rmol shared:

```python
#!/usr/bin/env python3
import logging
import operator
import os
import re
EXPORT_CLASSES = {"07": "Printer", "08": "Storage"}
EXPORT_INTERFACE_CLASSES = EXPORT_CLASSES.keys()
VENDOR_ID_RE = re.compile("^(?P<id>[0-9A-Fa-f]{4})  (?P<name>.*)$")
PRODUCT_ID_RE = re.compile("^\t(?P<id>[0-9A-Fa-f]{4})  (?P<name>.*)$")
def load_product_info():
    info = {"vendors": {}, "products": {}}
    vendor_id = vendor_name = None
    product_id = product_name = None
    with open("/var/lib/usbutils/usb.ids") as f:
        for line in f:
            if line[0] == "#" or not line:
                continue
            p = PRODUCT_ID_RE.match(line)
            if p:
                product_id = p.group(1)
                product_name = p.group(2)
                info["products"][vendor_id][product_id] = product_name
            else:
                v = VENDOR_ID_RE.match(line)
                if v:
                    vendor_id = v.group(1)
                    vendor_name = v.group(2)
                    info["vendors"][vendor_id] = vendor_name
                    info["products"][vendor_id] = {}
                elif vendor_id:
                    # don't want all the class IDs and such below product info
                    break
    return info
def get_device_info(product_info, directory):
    interface_class_file = os.path.join(directory, "bInterfaceClass")
    try:
        if not os.path.isfile(interface_class_file):
            return None
        interface_class = open(interface_class_file).read().strip()
        if interface_class in EXPORT_INTERFACE_CLASSES:
            os.chdir(directory)
            parent = os.path.dirname(os.getcwd())
            bus_number = open(os.path.join(parent, "busnum")).read().strip()
            dev_number = open(os.path.join(parent, "devnum")).read().strip()
            vendor_id = open(os.path.join(parent, "idVendor")).read().strip()
            product_id = open(os.path.join(parent, "idProduct")).read().strip()
            info = {
                "bus": bus_number,
                "device": dev_number,
                "interface_class": EXPORT_CLASSES[interface_class],
                "product": product_info["products"].get(vendor_id, {}).get(product_id),
                "vendor": product_info["vendors"].get(vendor_id),
            }
            return info
    except Exception as e:
        logging.error("Could not read device info from %s: %s", directory, e)
def get_export_devices(product_info):
    export_devices = []
    for directory, subdirs, files in os.walk("/sys/bus/usb/devices"):
        for s in subdirs:
            s = os.path.join(directory, s)
            info = get_device_info(product_info, s)
            if info:
                export_devices.append(info)
    return export_devices
if __name__ == "__main__":
    product_info = load_product_info()
    export_devices = get_export_devices(product_info)
    export_devices.sort(key=operator.itemgetter("product"))
    export_devices.sort(key=operator.itemgetter("vendor"))
    export_devices.sort(key=operator.itemgetter("device"))
    export_devices.sort(key=operator.itemgetter("bus"))
    export_devices.sort(key=operator.itemgetter("interface_class"))
    for d in export_devices:
        print(
            "Bus: {bus:>2s} Device: {device:>2s} Type: {interface_class} "
            "Vendor: {vendor} Product: {product}".format(**d)
        )
```

I'm going to close this pr from my fork since I want to open it up for other people on the team to commit to the working branch and contribute directly.

creviera commented:
Reopening because:

* `lsusb -s [bus_id]` -> `lsblk -p -o KNAME /dev/sda1` more consistently works
  - if the partition is not /dev/sda1, then `check_luks_volume` will also fail, so this may need discussion 
* we are going an entirely different route for export (we don't have to concern ourselves with changing bus ids anymore :tada: john is currently working on a script to auto-attach and remove `--persistent` when provisioning)
* and...
> The reason we used lsusb here is because we wanted to check_usb_connected function to be able to detect both printers and block devices.
> 
> The command here will return 32: lsblk: /dev/sda1/ is not a block device. I see 2 options:
> 
> a) search for a more reliable solution that works for printers and drives
> b) we can separate the preflight checks for USB connected:
> i. For disks, using the method described here
> ii. For printers, using either lsusb or another method
> 
> what do you think?

The current code does not need to rely on lsusb for print because we use [lpinfo](https://github.com/freedomofpress/securedrop-export/blob/78bcbfe6d26f5456f68cdd7c7b30353eaa7bb4a1/securedrop_export/export.py#L304) and return `ERROR_PRINTER_NOT_FOUND` if that returns "" for the uri or `ERROR_PRINTER_URI` if calling lsusb fails.

There will be follow-up work to rearchitect export but for now, switching from using `lsusb

eloquence commented:
>  lsusb -s [bus_id] -> lsblk -p -o KNAME /dev/sda1 more consistently works 

I've not yet gotten that to work, only when using `/dev/sda` as the device name instead `/dev/sda1`.

Testing with a LUKS device, I don't see any device files or `lsblk` entries for `/dev/sda1` or a higher number, I just see `/dev/sda`. When the device is unlocked and mounted, I see an entry like this in the plain `lsblk` output

```
sda                                           8:16   1  14.4G  0 disk  
└─luks-69667749-86a2-49d5-956d-d754c7a98a9f 253:1    0  14.4G  0 crypt /media/erik/enc
```

(This is on Ubuntu, but the output on Qubes is essentially the same).

creviera commented:
Ah, yes, I remember you showing me that last time and our idea was to check for `sda` instead of `sda1`. Then when it comes time to decrypt, we would make sure a) that there is only one partition under `sda`, and b) check if it's luks encrypted. I'm asking @kushaldas for review and feedback. He's testing now.

kushaldas commented:
>Testing with a LUKS device, I don't see any device files or lsblk entries for /dev/sda1 or a higher number, I just see /dev/sda. When the device is unlocked and mounted, I see an entry like this in the plain lsblk output

This is because I think you have the whole device encrypted and then partitions inside of it. In other cases, we first create a partition and then encrypt it.

kushaldas commented:
> Ah, yes, I remember you showing me that last time and our idea was to check for `sda` instead of `sda1`. Then when it comes time to decrypt, we would make sure a) that there is only one partition under `sda`, and b) check if it's luks encrypted. I'm asking @kushaldas for review and feedback. He's testing now.

Just checking for `sda` will give enough details to tell us that it is a storage device, and in future if you can loop over the partitions to find the first encrypted partition and stop the loop and go with decrypt, mounting the partition ... regular steps.

This PR is a good way to handle the issue in my mind.

creviera commented:
> sda                                           8:16   1  14.4G  0 disk  
> └─luks-69667749-86a2-49d5-956d-d754c7a98a9f 253:1    0  14.4G  0 crypt /media/erik/enc

@eloquence, looks like you're using the default `lsblk` output:
```
NAME                    MAJ:MIN RM   SIZE RO TYPE  MOUNTPOINT
sda                       8:16   1  14.4G  0 disk  
└─luks-6966...          253:1    0  14.4G  0 crypt /media/erik/enc
```

If you do:
```
> sudo cryptsetup isLuks "sda"
> echo $?
```
do you get an output of 0?

emkll commented:
> Testing with a LUKS device, I don't see any device files or lsblk entries for /dev/sda1 or a higher number, I just see /dev/sda. When the device is unlocked and mounted, I see an entry like this in the plain lsblk output

When creating the export device, did you use MBR/DOS or EPT as the partitioning scheme? So far, I think we've all used the former (MBR/DOS) ? It would be interesting to see if this is what causes the inconsistency here.

I think the approach by @creviera of iterating over the partitions in sda makes sense.

> (This is on Ubuntu, but the output on Qubes is essentially the same).

@eloquence (and potentially others): I would strongly advise to test these scripts/tooling under qubes. Testing on multiple platforms (e.g. Ubuntu) introduces needless confusion to an already very complex flow. Some reasons come to mind:

1. We want to ensure everyone is running the similar/identical package versions
2. We want to ensure the system configuration is similar
3. the kernel version/config is similar/identical (for example, network device names can change depending on kernel configurations)
4. Especially when dealing with USB, having other devices plugged in may have side-effect (e.g. numbering)
5. When unlocking partitions, you might accidentally cause breakage to your system. Isolating it to a VM is probably best

eloquence commented:
[tested in Qubes/Debian VM]

> sudo cryptsetup isLuks "sda"
> echo $?

That command results in "Device sda doesn't exist or access denied.", exit code 4. However, the command

> sudo cryptsetup isLuks "/dev/sda"

results in exit code 0 (no message).

> When creating the export device, did you use MBR/DOS or EPT as the partitioning scheme? 

I don't recall; it was done through the GNOME Disk utility. I'll try to create a fresh one and write down the steps.

eloquence commented:
OK, the only way I've found to create a USB drive in `gnome-disks` that doesn't have a `/dev/sda1` partition is as follows:

1) Select the device
2) Select "Format disk" from the top right hamburger menu
3) Select "No partitioning (empty)" from the partitioning options
4) Click the "Gears" icon
5) Select "Format partition"
6) Create a LUKS volume.

This results in a device like `/dev/sda` without a partition number. It is otherwise perfectly usable and will be auto-detected in GNOME etc. (Using GPT or MBR instead of the "No partitioning" option in step 3 both result in a `/dev/sda1` partition.)

I suspect I created this particular LUKS volume using the path above.

creviera commented:
> > sudo cryptsetup isLuks "/dev/sda"

> results in exit code 0 (no message).

perfect, that's what I meant :]

I still need to resolve conflict but this should be ready for testing.

creviera commented:
rebased and updated tests. ready for review.

creviera commented:
Note: this implements part of our solution for https://github.com/freedomofpress/securedrop-export/issues/18 (we still need to remove persistent attachment from the Export VM and add the multiple-devices error, all of which will be done separately)

creviera commented:
## Test plan

This resolves the inconsistent bus ID ordering issue reported here: https://github.com/freedomofpress/securedrop-workstation/issues/307#issuecomment-533754121 causing the `lsusb` check to fail when the bus id is reported different between the export vm and dom0.

Both Erik and I have been able to see this issue frequently, so if you have problems repro'ing based on the information in that issue linked above, please let one of us know.

Also please test and see Export now works in Erik's case where `/dev/sda1` does not exist.
```
NAME                    MAJ:MIN RM   SIZE RO TYPE  MOUNTPOINT
sda                       8:16   1  14.4G  0 disk  
└─luks-6966...          253:1    0  14.4G  0 crypt /media/erik/enc
```

creviera commented:
> let's agree on and provide very high level documentation on what partitioning scheme, and which operating system we use to format and partition the drives

sgtm

creviera commented:
I addressed the requested change to update the docs to describe what kind of export devices we support and how to create them.

I updated both the luks-encryption check and code to mount the device in order to support as many ways of creating luks-encrypted devices as I could think of using `Disks` and added support for full-disk encryption with luks as well.

Test this PR with:
1. MBR-partitioned luks-encrypted device using `Disks`
2. GPT-partitioned luks-encrypted device using `Disks`
3. Full-disk luks encrypted device using `cryptsetup`
See docs for more instructions on how to prepare your device to be a supported export device.

emkll commented on securedrop_export/export.py:
```diff
@@ -164,76 +164,84 @@ def extract_tarball(self):
+        try:
+            # get the encrypted device name
+            self.set_extracted_device_name()
+            luks_header = subprocess.check_output(["sudo", "cryptsetup", "luksDump", self.device])
```
:+1: this will definitely improve the reliability of the luks checks. Since we shouldn't expect / support more than 1 luks volume this should work just fine.
